### ISSUE TITLE
Model Buildings.Fluid.HeatPumps.ModularReversible.Examples.AirToWater2D_OneRoomRadiator fails verification and has annotation issue

### ISSUE THREAD

**Issue by AndreaBartolini** [OpenModelica/OpenModelica#12710]
The following variables of the model `Buildings.Fluid.HeatPumps.ModularReversible.Examples.AirToWater2D_OneRoomRadiator` fail the verification vs the Dymola reference file:
- `heaPum.refCyc.sigBus.onOffMea`
- `heaPum.refCyc.sigBus.hea`
- `temRet.T`
- `temSup.T`
- `vol.T`
- `time`

Regardig the `time` variable the problem is that, for some reason, the `experiment` annotation is ignored by OpenModelica:
```modelica
  experiment(
      StartTime=0,
      StopTime=86400,
      Tolerance=1e-08)
```
and the simulation ends @ `time`=1, maybe the problem is related to the ticket #11794.

In any case the simulation on the following architecture confirms the verification failure for the above listed variables:
OpenModelica v1.24.0-dev-187-ge87692cb93 (64-bit) (MinGW - Clang)
Dymola Version 2024x Refresh 1, 2024-04-12 (64-bit) (VisualStudio 2022)
WIN 11 pro (64-bit)
Buildings master b57b61c4780a4e18d1e83df0fd7f839a7b4d3add

![image](https://github.com/OpenModelica/OpenModelica/assets/25974985/9831b2ca-b3c2-4868-9c6e-917c55a96fb8)

From the temperatures point of view the differences of the `heaPum.refCyc.sigBus.onOffMea` control variable in the first part of the transient seem to have less effect, the main effect is in the time interval of about [45200 - 52800] seconds.

There are two main causes that generate the differences in the control variables `heaPum.refCyc.sigBus.onOffMea` and  `heaPum.refCyc.sigBus.hea`: the thresholds with hysteresis `heaPum.hys` and `oneRooRadHeaPumCtr.hysSecCoo`.
The first has the low limit set to 1e-15, that is too small to ensure the trigger in the tolerance used for the simulation (see also https://github.com/lbl-srg/modelica-buildings/issues/3930 ), both of them have the threshold hysteresis that seems to be insufficient to avoid spurious interventions of the control (but I don't know the expected behavior of the control system designed for this model). 

In the following is shown the transient obtained by using the following settings:
- `heaPum.hys`: `uLow`=1e-4, `uHigh`=0.2,
- `oneRooRadHeaPumCtr.hysSecCoo`: `uLow`=`TRadMinSup` - 1, `uHigh`=`TRadMinSup` + 3,

![image](https://github.com/OpenModelica/OpenModelica/assets/25974985/32e8abd3-8e66-4458-929b-9a98e3f43c5a)

Obviously, maybe this is not the expected behavior of the control system designed for this model.

**Comment by AndreaBartolini** [comment]
The same analysis is also applicable to the following similar models:
- `Buildings.Fluid.HeatPumps.ModularReversible.Examples.CarnotWithLosses_OneRoomRadiator` 
- `Buildings.Fluid.HeatPumps.ModularReversible.Examples.Modular_OneRoomRadiator`

**Comment by AndreaBartolini** [comment]
The answer is not so simple...
After fixing eps-problem (only the lower bound), there are still deviations due to the following closed-loop logic:

![image](https://github.com/user-attachments/assets/5c8671b3-3e16-4b40-a50f-ad8bf9b16513)

in which the output `safCtr.yOut` depends on the output of the threshold `hys` and vice-versa, the algebraic loop is opened by the `preOnOff` block inside the `safCtr.onOffCtr` block but the circular dependency still remains and generates a complex event iteration @ time = 47804.61661290314 in OpenModelica and a less complex event iteration in Dymola. Unfortunately this two event iterations converge to a different solution:

![image](https://github.com/user-attachments/assets/90c6ea16-251e-438c-8bd7-94b770841ccf)

and this because the event iteration solver implemented in Dymola is slightly different from the one implemented in OpenModelica and in case of rush conditions that often are present in complex closed-loop logic some differences can arise.

By enlarge the bandwidth in the `oneRooRadHeaPumCtr.hysSecCoo` probably some rush condition is solved and the difference disappears, but the point is: how many enlarge? Difficult to answer.
The more robust solution could be to synchronize the logic in order to better control the event iteration, but this way is hard and probably foresees a deeply revision of the way to build the logic in the library (I think is quite impossible at the current stage of development).

In conclusion, the only way I see is to fine tuning the logic (in this case the bandwidth of  `oneRooRadHeaPumCtr.hysSecCoo`) to align the results between OpenModelica and Dymola, I think it is no possible to find criteria to put some warning in OpenModelica.

**Comment by FWuellhorst** [comment]
Ok, thanks for the clarification and your work! 
I think the algebraic loop is only required as the on/off-switch of the heat pump depends on whether it was previously turned on or off. Do you know what the is status on state machines for OM? From my point of view this is a classic problem which state machines would solve.

**Comment by AndreaBartolini** [comment]
> Do you know what the is status on state machines for OM?
As far as I know the state machines work in OM, @casella can you confirm?

**Comment by casella** [comment]
@casella and @AndreaBartolini will have a meeting and try to come up with some MWE that clearly show the issue. We need to clearly identify whether we have
- a tool issue
- a library issue
- a language specification issue

**Comment by casella** [comment]
> > Do you know what the is status on state machines for OM?
> 
> As far as I know the state machines work in OM, @casella can you confirm?

Not sure what is the status with the new backend.

**Comment by AndreaBartolini** [comment]
From the Buildings library point of view, the PR https://github.com/lbl-srg/modelica-buildings/pull/3949 solves the issue, so I close the ticket.

@casella maybe a new ticket about https://github.com/OpenModelica/OpenModelica/issues/12710#issuecomment-2238574651 should be opened...

**Comment by casella** [comment]
> [@casella](https://github.com/casella) maybe a new ticket about [#12710 (comment)](https://github.com/OpenModelica/OpenModelica/issues/12710#issuecomment-2238574651) should be opened...

Feel free to open one 😃
